### PR TITLE
kernel: get rid of multiple global variables

### PR DESCRIPTION
**Pull request by fingolfin**
We are using plists as stacks in various places in the code. Strangely enough, we then also usually keep a separate stack pointer -- even in cases were the length of the plist is exactly equal to the value of the stack pointer. This seems pointless, and is a potential source of bugs, if the two values get out of sync by accident. It also increases the complexity of the code, making it slightly harder to understand ("hmm, it seems this variable is supposed to always contain the length of that plist -- but I can't be sure, so I'll now have to check the whole kernel to verify").

This PR builds upon PR #1605, which therefore should be merged first. (And afterwards, I could split this PR into several PRs which are easier to review; let me know if you would prefer that).

### CODE REVIEW

File: src/plist.h [hunk 1]
```diff
@@ -238,8 +238,29 @@ extern void            AssPlist (
...
+**  instead.
+**
+*/
+static inline UInt ADD_PLIST(Obj list, Obj val)
+{
+    const UInt pos = LEN_PLIST(list) + 1;
+    GROW_PLIST( list, pos );
+    SET_LEN_PLIST( list, pos );
+    SET_ELM_PLIST( list, pos, val );
+    if (IS_BAG_REF(val))
```
Comment (ChrisJefferson) [inline src/plist.h]: We don't do this anywhere else -- if we want to start doing this, I'd prefer it be in a separate commit, maybe even with a helper function of some kind -- `CHANGED_IN_BAG(list, val)` (picking a horrible random name)

File: src/intrprtr.c [hunk 2]
```diff
@@ -924,19 +896,18 @@ void            IntrWhileEnd ( void )
...
     STATE(IntrCoding) = 0;
     CodeEnd( 0 );
 
     /* If we are in a break loop, then we will have created a "dummy" local
        variable names list to get the counts right. Remove it */
-    if (STATE(CountNams) > 0)
-      STATE(CountNams)--;
+    const UInt len = LEN_PLIST(STATE(StackNams));
+    if (len > 0)
+      SET_LEN_PLIST(STATE(StackNams), len - 1);
```
Comment (ChrisJefferson) [inline src/intrprtr.c]: I worry about not zeroing out the removed value here?

Comment (fingolfin) [inline src/intrprtr.c]: Not sure what the worry is? Note that the code never zeroed out the old value.

If we zero them, this would cause a deeper change in semantics, as now those objects referenced before we zeroed out the ref could be garbage collected. I *think* that should be safe, but it's still a risk, so I felt it best to leave that to a future PR.

File: src/plist.h (hunk 1 above)
Comment (fingolfin) [inline src/plist.h]: Well, the reason I put this in here is that some code that I changed to use `ADD_PLIST` was clever enough to not invoked `CHANGED_BAG` when inserting intobjs or NULL pointers. So I figured this was safe optimization.

I am not sure what the helper function would gain, other than being yet another function you'd have to know by heart in order to read kernel code ;-). Perhaps if it did something more elaborate than the two lines it'd replace here?

Comment (fingolfin) [inline src/plist.h]: Anyway, I'll remove this optimization -- but I'll do so in PR #1605, where this change logically belongs. Once that PR has been merged, I can then rebase this PR.

File: src/intrprtr.c (hunk 2 above)
Comment (ChrisJefferson) [inline src/intrprtr.c]: You are right, let's just leave the semantics the same as before.

File: src/plist.h (hunk 1 above)
Comment (ChrisJefferson) [inline src/plist.h]: I can't see anywhere else in the GAP source code where we skip a `CHANGED_BAG` by checking if the thing we added is itself a bag or not (maybe we do somewhere), I had to convince myself this was correct behaviour.

We could make such a change, but we should probably make it to lots of other methods (`AssPlist` being an obvious choice).

Comment (fingolfin) [inline src/plist.h]: I agree -- as I said, I'll remove this for now

File: src/intrprtr.c (hunk 2 above)
Comment (stevelinton) [inline src/intrprtr.c]: As discussed elsewhere, good practice is to zero out the unused entries in this situation, but it should be done explicitly when it is known to be correct.
